relay/network_graph/dijkstra_weighted.py

```diff
@@ -111,7 +46,81 @@ def find_path(G, source, target, get_fee, value, max_hops=None, max_fees=None):
                 push(fringe, (n+1, vu_dist, next(c), u))
                 paths[u] = paths[v] + [u]
     try:
-        return (dist[target]-value, paths[target])
+        return (dist[target]-value, paths[target])  # cost is the total fee, not the actual amount to be transfered
     except KeyError:
         raise nx.NetworkXNoPath(
             "node %s not reachable from %s" % (source, target))
+
+
+def find_path_triangulation(G, source, target_reduce, target_increase, get_fee, value, max_hops=None, max_fees=None):
+    # target_reduce is the node we want to reduce our credit with (for value > 0)
+    # target_increase is the node which will result with an increased debt (for value > 0)
+    # must return a path and the total fee of that path
+    # to be called in the right order with source as source and target as target, contrary to find_path
+
+    # value may be negative as far as I know
+    # value < 0 means source wants to increase the amount target_reduce owes her
+        # (give negative value to target_reduce)
+
+    G_succ = G.succ if G.is_directed() else G.adj
+
+    neighbors = [x[0] for x in G_succ[source].items()]
+
+    if target_reduce not in neighbors:
+        raise nx.NetworkXNoPath(
+            "node %s not a neighbor of %s" % (target_reduce, source))
+
+    if target_increase not in neighbors:
+        raise nx.NetworkXNoPath(
+            "node %s not a neighbor of %s" % (target_increase, source))
+
+    if value >= 0:
```

Comment:
If you think you can not reduce the code, then I would say those two cases should go to different functions find_path_triangulation_reduce and find_path_triangulation_increase. It only makes sense to combine them if there share something.

Reply:
But I am sure it is possible to do it in one function if you find the similarities in the code and all the a < b if else things in just one line, after fixing the get fee or write your own get_fee function.